src/ChannelServer/World/Character.cs

```diff
@@ -555,6 +639,392 @@ public int MaxMAtk
+				//if(hasBuff("Finestra"))
+				//	value += 40 + ((SkillLevel - 1) * 15)
+
+				// Buffs: "BLK_BM"
+				var byBuffs = 0;
+
+				// Rate buffs:
+				var rate = 0;
+				var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs + byRateBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns Block Penetration.
+		/// </summary>
+		public int BlockPen
+		{
+			get
+			{
+				var Level = this.Level;
+				var stat = this.Dex;
+
+				var byLevel = Level / 4f;
+				var byStat = (stat / 2f) + ((float)Math.Floor(stat / 15f) * 3f);
+				var byItem = 0; // TODO: Cached Block Penetration for inventory/equip ("BLK_BREAK")
+
+				var value = byLevel + byStat + byItem;
+
+				// Buffs: "BLK_BREAK_BM"
+				var byBuffs = 0;
+
+				// Rate buffs: 
+				var rate = 0;
+				var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs + byRateBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns Aoe Attack Ratio.
+		/// </summary>
+		public int AoeRatio
+		{
+			get
+			{
+				var baseValue = 4;
+				var byItem = 0f; // TODO: Cached Aoe Attack Ratio for inventory/equip  ("SR")
+
+				var value = baseValue + byItem;
+
+				//if(hasBuff("Finestra") && hasAttribute("Finestra: Splash"))
+				//	value += 3
+
+				// Buffs: "SR_BM"
+				var byBuffs = 0;
+
+				// Rate buffs: TOS don't have a Aoe attack Ratio rate bonus (correct me if im wrong)
+				//var rate = 0;
+				//var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns Aoe Defence Ratio.
+		/// </summary>
+		public int AoeRatioDef
+		{
+			get
+			{
+				var baseValue = 1;
+				var byItem = 0f; // TODO: Cached Aoe Defence Ratio for inventory/equip ("SDR")
+
+				var value = baseValue + byItem;
+
+				// Buffs: "SDR_BM"
+				var byBuffs = 0;
+
+				// Rate buffs: TOS don't have a Aoe Defence Ratio rate bonus (correct me if im wrong)
+				//var rate = 0;
+				//var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns HP Recovery amount.
+		/// </summary>
+		public int RHp
+		{
+			get
+			{
+				var baseValue = 1;
+				var level = this.Level;
+				var stat = this.Con;
+
+				var ByLevel = level / 2f;
+				var byStat = (stat * 2f) + ((float)Math.Floor(stat / 5f) * 3f);
+				var byItem = 0f; // TODO: Cached HP Recovery for inventory/equip ("RHP")
+
+				var value = baseValue + ByLevel + byStat + byItem;
+
+				// Buffs: "HSP_BM"
+				var byBuffs = 0;
+
+				// Rate buffs: 
+				var rate = 0;
+				var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns SP Recovery amount.
+		/// </summary>
+		public int RSp
+		{
+			get 
+			{
+				var baseValue = 3;
+				var level = this.Level;
+				var stat = this.Spr;
+
+				var ByLevel = level / 2f;
```

Comment:
I think this is incorrect, the client files have `local byLevel = lv * jobObj.JobRate_SP;` for `byLevel`.